spec/presenters/idv/in_person/verification_results_email_presenter_spec.rb

```diff
@@ -29,4 +33,133 @@
+    context 'without service provider' do
+      let(:sp) { nil }
+
+      it 'returns app name' do
+        expect(presenter.service_provider_or_app_name).to eq(APP_NAME)
+      end
+    end
+
+    context 'with service provider' do
+      let(:sp) { create(:service_provider) }
+
+      it 'returns friendly name of service provider' do
+        expect(presenter.service_provider_or_app_name).to eq(sp.friendly_name)
+      end
+    end
+  end
+
+  describe '#show_cta?' do
+    context 'without service provider' do
+      let(:sp) { nil }
+
+      it { expect(presenter.show_cta?).to eq(true) }
+    end
+
+    context 'with service provider' do
+      let(:homepage_url) { nil }
+      let(:sp) { create(:service_provider) }
+
+      before do
+        resolver = instance_double(SpReturnUrlResolver)
+        allow(resolver).to receive(:homepage_url).and_return(homepage_url)
+        allow(SpReturnUrlResolver).to receive(:new).with(service_provider: sp).and_return(resolver)
+      end
+
+      context 'without homepage_url' do
+        let(:homepage_url) { nil }
+
+        it { expect(presenter.show_cta?).to eq(false) }
+      end
+
+      context 'with homepage_url' do
+        let(:homepage_url) { 'https://example.com' }
+
+        it { expect(presenter.show_cta?).to eq(true) }
+      end
+    end
+  end
+
+  describe '#sign_in_url' do
+    context 'without service provider' do
+      let(:sp) { nil }
+
+      it 'returns root url' do
+        expect(presenter.sign_in_url).to eq(root_url)
+      end
+    end
+
+    context 'with service provider' do
+      let(:homepage_url) { nil }
+      let(:sp) { create(:service_provider) }
+
+      before do
+        resolver = instance_double(SpReturnUrlResolver)
+        allow(resolver).to receive(:homepage_url).and_return(homepage_url)
+        allow(SpReturnUrlResolver).to receive(:new).with(service_provider: sp).and_return(resolver)
+      end
+
+      context 'without homepage_url' do
+        let(:homepage_url) { nil }
+
+        it 'returns root url' do
+          expect(presenter.sign_in_url).to eq(root_url)
+        end
+      end
+
+      context 'with homepage_url' do
+        let(:homepage_url) { 'https://example.com' }
+
+        it 'returns homepage url' do
+          expect(presenter.sign_in_url).to eq(homepage_url)
+        end
+      end
+    end
+  end
+
+  describe '#service_provider_homepage_url' do
+    context 'without service provider' do
+      let(:sp) { nil }
+
+      it 'returns nil' do
+        expect(presenter.service_provider_homepage_url).to be_nil
+      end
+    end
+
+    context 'with service provider' do
+      let(:homepage_url) { nil }
+      let(:sp) { create(:service_provider) }
+
+      before do
+        resolver = instance_double(SpReturnUrlResolver)
+        allow(resolver).to receive(:homepage_url).and_return(homepage_url)
+        allow(SpReturnUrlResolver).to receive(:new).with(service_provider: sp).and_return(resolver)
+      end
```

Comment:
in the spirit of "don't stub what you don't own" I'd prefer if there was a `sp_return_url_resolver` method on the presenter that we stubbed here instead of stubbing new

Reply:
> in the spirit of "don't stub what you don't own" I'd prefer if there was a `sp_return_url_resolver` method on the presenter that we stubbed here instead of stubbing new

Some day I hope to do this correctly on the first pass 😅

Reply:
> in the spirit of "don't stub what you don't own" I'd prefer if there was a `sp_return_url_resolver` method on the presenter that we stubbed here instead of stubbing new

Updated in 1d43d4449.